doc/source/optimization.md

```diff
@@ -1,67 +1,264 @@
+    the new image. The name hook-image-puller is a technical name referring to
+    how a [Helm hook](https://docs.helm.sh/developing_charts/#hooks) is used to
+    accomplish this, a more informative name would had been
+    *pre-upgrade-image-puller*.
 
-```yaml
-prePuller:
-  continuous:
-    enabled: true
-```
+    The hook-image-puller is enabled by default. To disable it, use the
+    following snippet in your `config.yaml`:
+
+    ```yaml
+    prePuller:
+      hook:
+        enabled: false
+    ```
+
+2. A node is added (Cluster Autoscaler)
 
-### Pre-pulling additional images
+    The amount of nodes in a Kubernetes cluster can increase, either by manually
+    scaling up the cluster size or by a cluster autoscaler. As new nodes will
+    come fresh without any images on their disks, a user pod arriving to this
+    node will be forced to wait while the image is pulled.
 
-By default, the pre-puller only pulls the singleuser image & the networktools
-image (if access to cloud metadata is disabled). If you have customizations that
-need additional images present on all nodes, you can ask the pre-puller to also
-pull an arbitrary number of additional images.
+    With the *continuous-image-puller* enabled (disabled by default), the user's
+    container image will be pulled when a new node is added. New nodes can for
+    example be added manually or by a cluster autoscaler. The continuous
+    image-puller uses a
+    [daemonset](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/)
+    to force Kubernetes to pull the user image on all nodes as soon as a node is
+    present.
+
+    The continuous-image-puller is disabled by default. To enable it, use the
+    following snippet in your `config.yaml`:
+
+    ```yaml
+    prePuller:
+      continuous:
+        # NOTE: if used with a Cluster Autoscaler, also add user-placeholders
+        enabled: true
+    ```
+
+    It is important to realize that if the continuous-image-puller together with
+    a Cluster Autoscaler (CA) won't guarantee a reduced wait time for users. It
+    only helps if the CA scales up before real users arrive, but the CA will
+    generally fail to do so. This is because it will only add a node if one or
+    more pods won't fit on the current nodes but would fit more if a node is
+    added, but at that point users are already waiting. To scale up nodes ahead
+    of time we can use *user-placeholders*.
+
+### Configuring the pulled images
+
+The hook-image-puller and the continuous-image-puller has various sources to
+know what images it should pull. These sources are all found in the values
+provided with the Helm chart (that can be overridden with `config.yaml`) under
+the following paths:
+
+#### Relevant image sources
+- `singleuser.image`
+- `singleuser.profileList[].kubespawner_override.image`
+- `singleuser.extraContainers[].image`
+- `prePuller.extraImages[].someName`
```

Comment:
Some of these are not explained elsewhere in the z2jh guide, right? E.g. `extraImages` means arbitrary other images that get pulled (and are run?) on a node?

Reply:
They may be yes, but hmmm. What I want to convey in this section is not about what profileList does etc, but simply that the image-pullers is aware of images in these configuration options and will pull them as well by default.

I'll look through this section again.